system/lib/wasmfs/virtual.h

```diff
@@ -0,0 +1,116 @@
+// Copyright 2022 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// Utilities for creating virtual backends that forward operations to underlying
+// "real" files.
+
+#pragma once
+
+#include "file.h"
+
+namespace wasmfs {
+
+static inline std::shared_ptr<File> devirtualize(std::shared_ptr<File>);
+
+class VirtualDataFile : public DataFile {
+protected:
+  std::shared_ptr<DataFile> real;
+
+public:
+  VirtualDataFile(std::shared_ptr<DataFile> real, backend_t backend)
```

Comment:
Should it be protected? Unless there is a plan to instantiate it.

Reply:
No, unfortunately the constructors need to be public so that they can be called by `std::make_shared`. The usual alternative, `friend`, doesn't work here, because it would require specifying the internal library function that actually calls the constructor, and that's an implementation detail that could change.